error.go

```diff
@@ -88,24 +89,34 @@ func (e *wrappedManyError) Is(target error) bool {
 	return false
 }
 
+func (e *wrappedManyError) As(target interface{}) bool {
+	for _, err := range e.errs {
+		if errors.As(err, target) {
+			return true
+		}
+	}
+
+	return false
+}
+
 func wrapAll(errs ...error) error {
```

Comment:
What's `wrapAll` used for? We need comments for private APIs.

Reply:
`wrapAll` and `wrapManyError` are used to return multiple errors. For when an error occurs while cleaning up in response to a previous error. I'm not positive this is a good idea.

Reply:
It is also used to combine multiple errors when more than one connection in a pool errors on connect or close. Again I'm not sure this is better than just returning the first error.

Reply:
Why not use the standard go error chaining approach (with the Unwrap method)? Does it have to be a multi-error?

Reply:
where `wrapAll()` is used we have no idea if the errors are duplicates of the same error or if they are different errors. Putting multiple errors into one chain is 
1. confusing `Unwrap()` is a sinonym for `Cause()`, but the two errors are not causally related so how would you flatten them into a single chain?
2. problematic to implement: to make one error return another when it is unwrapped you must know how the error is implemented which is usually hidden behind private fields/methods. Also, not all error implementations implement unwrap. Also, sticking an error on the the end of a chain changes the equality of at least some of the errors in the chain so it would alter `errors.Is()` behavior for some errors in the chain.

Reply:
The more I think about this the more I think its better to always return the first error :thinking:

Reply:
There is one scenario I'm not sure about on always returning the first error. Can the server return multiple errors in response to one query? I vaguely remember that happening but I'm not positive. 

`wrapAll()` is used to accumulate errors when the server returns more than one. Not sure if its a good idea to only return the first error in this case.

Reply:
> There is one scenario I'm not sure about on always returning the first error. Can the server return multiple errors in response to one query? I vaguely remember that happening but I'm not positive.

I don't think there's a scenario like that, to be honest. The server can only return one error per operation. Now, while you process the server error response you could have other errors generated by the client code, but that should be exceptionally rare and is a bug that we'll need to fix.

In general, we should return server errors. If while handling a server error E1 we generate a client-side error E2 because of some bug on our side, we should return E2 wrapping E1.

> wrapAll() is used to accumulate errors when the server returns more than one. Not sure if its a good idea to only return the first error in this case.

I don't think it's that useful, TBH. About the only use case would be the `Connect()` function, in which you might resolve a hostname to a bunch of IP addresses, try them all, fail on all of them, and then aggregate all errors into one. But in Go's case I don't think that's useful. So yeah, I'd just drop the concept of multi-error.

Reply:
> If while handling a server error E1 we generate a client-side error E2 because of some bug on our side, we should return E2 wrapping E1.

I'm not sure this makes sense semantically. Error chaining in go is a way of adding context to the root error. Think of it like building up something like a stack trace. For example you got an `I/O Timeout` _while_ reading from a socket connected to `tcp address 1.2.3.4:80` _while_ processing an RPC call _while_ logging `user x` into your system.

The E1/E2 example has an `AND` in it. I got a query error `AND` then a network error while reading the rest of the messages from the server _while_ executing a query. Chaining these together makes E2 the _cause_ of E1. I don't know if that makes sense. Maybe its fine, I'm uncertain though.

@tailhook what do you do in Rust when you encounter an error you are going to return and get a second error while cleaning up from the first?

Reply:
To clarify, I think the `wrapAll()` bit should be removed. The question is what to do with errors while cleaning up from a current error.